File: python-algorithm/ds_9_bit_operation/L137-m.py
# -*- coding: utf-8 -*-


# 1 使用哈希标记次数



# 2 位操作，太难想到了
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        seenOnce = 0
        seenTwice = 0
        for num in  nums:
            seenOnce = ~seenTwice & (seenOnce ^ num)
            seenTwice = ~seenOnce & (seenTwice ^ num)
        return seenOnce;

File: python-algorithm/ds_9_bit_operation/L137-m.cpp

#include <iostream>
#include <vector>

using namespace std;

//很特殊的技巧，还没弄懂
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int seenOnce = 0, seenTwice = 0;

        for (int num : nums) {
            // first appearence:
            // add num to seen_once
            // don't add to seen_twice because of presence in seen_once

            // second appearance:
            // remove num from seen_once
            // add num to seen_twice

            // third appearance:
            // don't add to seen_once because of presence in seen_twice
            // remove num from seen_twice
            seenOnce = ~seenTwice & (seenOnce ^ num);
            seenTwice = ~seenOnce & (seenTwice ^ num);
        }

        return seenOnce;
    }
};